viewer/src/datamodels/cad/MaterialManager.ts

```diff
@@ -141,61 +68,55 @@ export class MaterialManager {
+    this.updateNodeAppearance(appearanceProvider, materials, treeIndices);
   }
 
-  private updateRenderEffectBuffers(
-    renderAppearance: ModelRenderAppearance | undefined,
+  private updateNodeAppearance(
+    appearanceProvider: NodeAppearanceProvider | undefined,
     materials: Materials,
     treeIndices: number[]
   ) {
-    if (!renderAppearance || renderAppearance.renderInFront === undefined) {
+    if (!appearanceProvider) {
       return;
     }
 
-    for (const treeIndex of treeIndices) {
-      materials.overrideVisibilityPerTreeIndex.image.data[4 * treeIndex + 1] = renderAppearance.renderInFront(treeIndex)
-        ? 255
-        : 0;
-    }
-    materials.overrideVisibilityPerTreeIndex.needsUpdate = true;
-  }
-
-  private updateGlobalNodeAppearance(modelIdentifier: string, materials: Materials, treeIndices: number[]) {
-    if (this._globalAppearance && this._globalAppearance.color !== undefined) {
-      updateGlobalColors(modelIdentifier, this._globalAppearance.color, materials, treeIndices);
-      materials.overrideColorPerTreeIndex.needsUpdate = true;
-    }
-    if (this._globalAppearance && this._globalAppearance.visible !== undefined) {
-      updateGlobalVisibility(modelIdentifier, this._globalAppearance.visible, materials, treeIndices);
-      materials.overrideVisibilityPerTreeIndex.needsUpdate = true;
-    }
-  }
-
-  private updateLocalNodeAppearance(
-    localAppearance: ModelNodeAppearance | undefined,
-    materials: Materials,
-    treeIndices: number[]
-  ) {
-    if (localAppearance && localAppearance.color !== undefined) {
-      updateColors(localAppearance.color, materials, treeIndices);
-      materials.overrideColorPerTreeIndex.needsUpdate = true;
-    }
-    if (localAppearance && localAppearance.visible !== undefined) {
-      updateVisibility(localAppearance.visible, materials, treeIndices);
-      materials.overrideVisibilityPerTreeIndex.needsUpdate = true;
+    const count = treeIndices.length;
+    for (let i = 0; i < count; ++i) {
+      const treeIndex = treeIndices[i];
+      const style = appearanceProvider.styleNode(treeIndex);
+
+      // Override color
+      if (style && style.color !== undefined) {
+        materials.overrideColorPerTreeIndex.image.data[4 * treeIndex] = style.color[0];
+        materials.overrideColorPerTreeIndex.image.data[4 * treeIndex + 1] = style.color[1];
+        materials.overrideColorPerTreeIndex.image.data[4 * treeIndex + 2] = style.color[2];
+        materials.overrideColorPerTreeIndex.image.data[4 * treeIndex + 3] = style.color[3];
+        materials.overrideColorPerTreeIndex.needsUpdate = true;
+      }
+
+      // Hide node?
+      if (style && style.visible !== undefined) {
+        materials.overrideVisibilityPerTreeIndex.image.data[4 * treeIndex] = style.visible ? 255 : 0;
+        materials.overrideVisibilityPerTreeIndex.needsUpdate = true;
+      }
+
+      // Render in front of everything (i.e. skip depth testing)?
```

Comment:
We don't skip depth testing, we clear the depth buffer between passes; but depth test in the highlighting pass (else the ordering between highlighted objects would be wack)

Reply:
good point